snuba/query/nodes.py

```diff
@@ -0,0 +1,189 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from enum import Enum
+from typing import Callable, Iterable, Optional, Sequence, Union
+
+
+class Node(ABC):
+    """
+    Abstract representation of anything iterable on a query.
+    The query designed as a tree. The root of the tree is the Query class.
+    The query class has several branches, which are almost all collections.
+    These branches represent selected_columns, conditions, etc.
+    All the elements of all those collections are subclasses of Node.
+    """
+    @abstractmethod
+    def format(self) -> str:
+        """
+        Turn this node into a string. for Clickhouse
+        TODO: provide a clickhouse formatter to this method so that, through
+        a strategy pattern, this class will provide the content for the query
+        and the clickhouse formatter will provide the format.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def map(self, closure: Callable[[Node], Node]) -> Node:
```

Comment:
map and filter are a bit of a bet. I think they can be very flexible as a model to edit parts of the query and they can be extended easily to a model where we can do pattern matching of subtrees and replacing them.  On the other end the typing may not be optimal.